src/handlers/notify_zulip.rs

```diff
@@ -1,96 +1,171 @@
+    github::{Issue, IssuesAction, IssuesEvent, Label},
     handlers::Context,
 };
 
 pub(super) struct NotifyZulipInput {
     notification_type: NotificationType,
+    /// Label that triggered this notification.
+    ///
+    /// For example, if an `I-prioritize` issue is closed,
+    /// this field will be `I-prioritize`.
+    label: Label,
 }
 
 pub(super) enum NotificationType {
     Labeled,
     Unlabeled,
+    Closed,
+    Reopened,
 }
 
 pub(super) fn parse_input(
     _ctx: &Context,
     event: &IssuesEvent,
     config: Option<&NotifyZulipConfig>,
-) -> Result<Option<NotifyZulipInput>, String> {
-    if let IssuesAction::Labeled | IssuesAction::Unlabeled = event.action {
-        let applied_label = &event.label.as_ref().expect("label").name;
-        if let Some(config) = config.and_then(|c| c.labels.get(applied_label)) {
-            for label in &config.required_labels {
-                let pattern = match glob::Pattern::new(label) {
-                    Ok(pattern) => pattern,
-                    Err(err) => {
-                        log::error!("Invalid glob pattern: {}", err);
-                        continue;
-                    }
-                };
-                if !event
-                    .issue
-                    .labels()
-                    .iter()
-                    .any(|l| pattern.matches(&l.name))
-                {
-                    // Issue misses a required label, ignore this event
-                    return Ok(None);
+) -> Result<Option<Vec<NotifyZulipInput>>, String> {
+    let config = match config {
+        Some(config) => config,
+        None => return Ok(None),
+    };
+
+    match event.action {
+        IssuesAction::Labeled | IssuesAction::Unlabeled => {
+            let applied_label = event.label.as_ref().expect("label").clone();
+            Ok(config
+                .labels
+                .get(&applied_label.name)
+                .and_then(|label_config| {
+                    parse_label_change_input(event, applied_label, label_config)
+                })
+                .map(|input| vec![input]))
+        }
+        IssuesAction::Closed | IssuesAction::Reopened => {
+            Ok(Some(parse_close_reopen_input(event, config)))
+        }
+        _ => Ok(None),
+    }
+}
+
+fn parse_label_change_input(
+    event: &IssuesEvent,
+    label: Label,
+    config: &NotifyZulipLabelConfig,
+) -> Option<NotifyZulipInput> {
+    if !has_all_required_labels(&event.issue, config) {
+        // Issue misses a required label, ignore this event
+        return None;
+    }
+
+    match event.action {
+        IssuesAction::Labeled if config.message_on_add.is_some() => Some(NotifyZulipInput {
+            notification_type: NotificationType::Labeled,
+            label,
+        }),
+        IssuesAction::Unlabeled if config.message_on_remove.is_some() => Some(NotifyZulipInput {
+            notification_type: NotificationType::Unlabeled,
+            label,
+        }),
```

Comment:
This previously looked like:

```rust
            if event.action == IssuesAction::Labeled && config.message_on_add.is_some() {
                return Ok(Some(NotifyZulipInput {
                    notification_type: NotificationType::Labeled,
                }));
            } else if config.message_on_remove.is_some() {
                return Ok(Some(NotifyZulipInput {
                    notification_type: NotificationType::Unlabeled,
                }));
            }
```

(notice the catch-all `else if`)

The new version should hopefully be clearer.